gnu-efi: 3.0.18 -> 4.0.0, adopt, modernise; gnu-efi_3: init at 3.0.19

lzcunt opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Note that this needs extensive testing of packages that depend on gnu-efi, as this update breaks ABI and may cause build failures and runtime errors. I've fixed building refind but I haven't tested running refind yet.

~~**Open question:** Should efilib license be added to lib/licenses.nix or should GNU-EFI use BSD-2-Clause instead? The efilib license is an obscure variant of the BSD-2-Clause license with an extra disclaimer about the EFI spec and says anything is "subject to change without notice".~~

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

lzcunt committed:
lib/licenses: add bsdAxisNoDisclaimerUnmodified

This is a variant composed of clause 1 and 3 of a BSD-Modified found in the
Linux kernel and this extra variant has the words "without modification" added
at the end of clause 1 either with or without a comma. It also lacks the
warranty disclaimer.

See <https://scancode-licensedb.aboutcode.org/bsd-no-disclaimer-unmodified.html>.

lzcunt committed:
gnu-efi: 3.0.18 -> 4.0.0, fix meta, format

This update may break packages that depend on gnu-efi, as this is an ABI
break and most upstream packages have not been updated to support this
release.

lzcunt committed:
gnu-efi: move to by-name/

lzcunt committed:
refind: fix build by using `gnu-efi_3` instead of `gnu-efi`

lzcunt commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 388761`

Logs: https://github.com/lzcunt/nixpkgs-review-gha/actions/runs/13771084773


---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>libsForQt5.discover</li>
    <li>plasma5Packages.discover</li>
    <li>qubes-core-vchan-xen</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>htcondor</li>
    <li>lanzaboote-tool</li>
    <li>python313Packages.guestfs</li>
    <li>python313Packages.guestfs.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 102 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>bootiso</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>distrobuilder</li>
    <li>docker-machine-kvm2</li>
    <li>efitools</li>
    <li>fdroidserver</li>
    <li>fdroidserver.dist</li>
    <li>fwupd</li>
    <li>fwupd-efi</li>
    <li>fwupd.debug</li>
    <li>fwupd.dev</li>
    <li>fwupd.devdoc</li>
    <li>fwupd.installedTests</li>
    <li>fwupd.lib</li>
    <li>fwupd.man</li>
    <li>gnome-boxes</li>
    <li>gnome-firmware</li>
    <li>gnome-software</li>
    <li>gnu-efi</li>
    <li>ipxe</li>
    <li>kdePackages.discover</li>
    <li>kdePackages.discover.debug</li>
    <li>kdePackages.discover.dev</li>
    <li>kdePackages.discover.devtools</li>
    <li>libguestfs</li>
    <li>libguestfs.guestfsd</li>
    <li>librenms</li>
    <li>libsForQt5.kinfocenter (plasma5Packages.kinfocenter)</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>libvmi</li>
    <li>libvmi.dev</li>
    <li>libvmi.lib</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>multipass</li>
    <li>nixpkgs-manual</li>
    <li>pantheon.elementary-greeter</li>
    <li>pantheon.elementary-onboarding</li>
    <li>pantheon.elementary-settings-daemon</li>
    <li>pantheon.switchboard-plug-about</li>
    <li>pantheon.switchboard-plug-pantheon-shell</li>
    <li>pantheon.switchboard-plug-power</li>
    <li>pantheon.switchboard-plug-security-privacy</li>
    <li>pantheon.switchboard-with-plugs</li>
    <li>pantheon.wingpanel-applications-menu</li>
    <li>pantheon.wingpanel-indicator-power</li>
    <li>pantheon.wingpanel-quick-settings</li>
    <li>pantheon.wingpanel-with-indicators</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>perl540Packages.SysVirt</li>
    <li>perl540Packages.SysVirt.devdoc</li>
    <li>podman-bootc</li>
    <li>prometheus-libvirt-exporter</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>python312Packages.libvirt</li>
    <li>python312Packages.libvirt.dist</li>
    <li>xen (python312Packages.xen)</li>
    <li>xen.boot (python312Packages.xen.boot)</li>
    <li>xen.dev (python312Packages.xen.dev)</li>
    <li>xen.doc (python312Packages.xen.doc)</li>
    <li>xen.man (python312Packages.xen.man)</li>
    <li>python313Packages.libvirt</li>
    <li>python313Packages.libvirt.dist</li>
    <li>python313Packages.xen</li>
    <li>python313Packages.xen.boot</li>
    <li>python313Packages.xen.dev</li>
    <li>python313Packages.xen.doc</li>
    <li>python313Packages.xen.man</li>
    <li>qemu_xen</li>
    <li>qemu_xen.debug</li>
    <li>qemu_xen.doc</li>
    <li>qemu_xen.ga</li>
    <li>refind</li>
    <li>rubyPackages.ruby-libvirt (rubyPackages_3_3.ruby-libvirt)</li>
    <li>rubyPackages_3_1.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>rubyPackages_3_4.ruby-libvirt</li>
    <li>sbsigntool</li>
    <li>solo5</li>
    <li>solo5.debug</li>
    <li>syslinux</li>
    <li>unetbootin</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-top</li>
    <li>virt-v2v</li>
    <li>virt-viewer</li>
    <li>virtnbdbackup</li>
    <li>virtnbdbackup.dist</li>
    <li>wimlib</li>
    <li>woeusb</li>
    <li>xen-guest-agent</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>libsForQt5.discover</li>
    <li>plasma5Packages.discover</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>ipxe</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 31 packages built:</summary>
  <ul>
    <li>efitools</li>
    <li>fwupd</li>
    <li>fwupd-efi</li>
    <li>fwupd.debug</li>
    <li>fwupd.dev</li>
    <li>fwupd.devdoc</li>
    <li>fwupd.installedTests</li>
    <li>fwupd.lib</li>
    <li>fwupd.man</li>
    <li>gnome-firmware</li>
    <li>gnu-efi</li>
    <li>kdePackages.discover</li>
    <li>kdePackages.discover.debug</li>
    <li>kdePackages.discover.dev</li>
    <li>kdePackages.discover.devtools</li>
    <li>libsForQt5.kinfocenter (plasma5Packages.kinfocenter)</li>
    <li>nixpkgs-manual</li>
    <li>pantheon.elementary-greeter</li>
    <li>pantheon.elementary-onboarding</li>
    <li>pantheon.elementary-settings-daemon</li>
    <li>pantheon.switchboard-plug-about</li>
    <li>pantheon.switchboard-plug-pantheon-shell</li>
    <li>pantheon.switchboard-plug-power</li>
    <li>pantheon.switchboard-plug-security-privacy</li>
    <li>pantheon.switchboard-with-plugs</li>
    <li>pantheon.wingpanel-applications-menu</li>
    <li>pantheon.wingpanel-indicator-power</li>
    <li>pantheon.wingpanel-quick-settings</li>
    <li>pantheon.wingpanel-with-indicators</li>
    <li>refind</li>
    <li>sbsigntool</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

lzcunt commented:
IANAL but the extra disclaimer added in the [efilib license](https://github.com/ncroxon/gnu-efi/blob/master/licenses/LICENSE.efilib) to the original BSD-2-Clause license doesn't make it a different license (it only applies to the EFI spec and information on the original website efilib was hosted in, and not the actual software) so I've gone and reverted adding the efilib license to lib/licenses. However, we've discovered in Matrix that GNU-EFI is also licensed under "BSD-Axis Without Modification" but upstream has mistakenly labeled it "BSD-2-Clause".

alyssais commented:
Ack license.

lzcunt committed:
gnu-efi: add lzcunt as maintainer

JohnRTitor commented:
Failing builds should be investigated. 

@ofborg build lanzaboote-tool

lzcunt commented:
I've tested booting into refind on qemu for x86_64 and aarch64. It boots into the menu and can reboot, shutdown, reboot into "Computer Setup Utility"/firmware menu on both architectures. I didn't test booting an entire system, it'd be awesome if @chewblacka or a refind user (especially on aarch64) could review and test it too. No idea why aarch64 tests fail.

JohnRTitor reviewed (approved):
These packages are already broken on master and can be ignored.
```
htcondor
lanzaboote-tool
python313Packages.guestfs
```

JohnRTitor commented:
I use refind. I will test this later today.

JohnRTitor commented:
![IMG_20250315_231323](https://github.com/user-attachments/assets/7ec3e562-8b1d-4606-abea-d50e3339be5f)

Installed refind using refind-install script.

I was hit with a "Signature validation failed" error even though I made sure to manually sign the refind bootloader EFI using sbctl. After an hour of trying over and over again, it didn't work. So I resorted to disabling secure boot and reinstalled refind.

So nothing actually shows up in the boot menu as you can see, only the banner image (which I set previously) shows up.

Definitely broke something on x64-linux.

JohnRTitor commented:
Rebuilt my system with the latest nixos-unstable (without this PR patch), and the bootloader broke on me, refuses to even show me the generations, anyone's having the same problem? 😔

JohnRTitor commented:
Recovered my system but still... 

![IMG_20250316_155252](https://github.com/user-attachments/assets/de192cb8-8ff0-4445-9bd1-8441fcf5faba)

I think we are better off init-ing a `gnu-efi_3` until refind gets an update.

lzcunt committed:
gnu-efi_3: init at 3.0.19

GNU-EFI was updated to 4.0.0, however rEFInd does not support 4.0.0 yet,
so we have to keep GNU-EFI 3.

JohnRTitor commented on pkgs/by-name/gn/gnu-efi_3/package.nix:
```diff
@@ -0,0 +1,62 @@
+
+assert lib.assertMsg stdenv.hostPlatform.isEfi "gnu-efi may only be built for EFI platforms";
+
+stdenv.mkDerivation rec {
```
finalAttrs

JohnRTitor commented on pkgs/by-name/re/refind/package.nix:
Could you add me to the maintainers here?

JohnRTitor commented on pkgs/by-name/gn/gnu-efi/package.nix:
```diff
@@ -0,0 +1,69 @@
+
+assert lib.assertMsg stdenv.hostPlatform.isEfi "gnu-efi may only be built for EFI platforms";
+
+stdenv.mkDerivation rec {
```
finalAttrs

lzcunt committed:
refind: add johnrtitor as maintainer

Authored-By: John Titor <50095635+JohnRTitor@users.noreply.github.com>

JohnRTitor commented on pkgs/by-name/gn/gnu-efi/package.nix:
```diff
@@ -0,0 +1,69 @@
+  src = fetchFromGitHub {
+    owner = "ncroxon";
+    repo = "gnu-efi";
+    rev = finalAttrs.version;
```
```suggestion
    tag = finalAttrs.version;
```

lzcunt commented:
I think this is pretty much ready, after @JohnRTitor confirms rEFInd works with 3.0.19. Also, is `ipxe` on aarch64-linux broken on master too or is that a regression in this PR? If it's caused by this PR we can make it use `gnu-efi_3` too.

JohnRTitor commented on pkgs/by-name/gn/gnu-efi_3/package.nix:
I am usually not a fan of overrideAttrs, on a major upgrade, to downgrade it. If both drv's diverge too much or if we add patches to the v4, it could v3, as it inherits them as well. But I suppose it's fine for now.

JohnRTitor reviewed (commented):
LGTM

JohnRTitor commented:
> is `ipxe` on aarch64-linux broken on master too or is that a regression in this PR? If it's caused by this PR we can make it use `gnu-efi_3` too.

oops, I merged it too fast :)

![image](https://github.com/user-attachments/assets/46aa6953-79da-407e-b327-d5777eac6265)

I think pinning should be the best option to fix the build.

JohnRTitor commented:
Or we could check if a "master" commit (https://github.com/ipxe/ipxe/tree/master) works. Needs testing though.

lzcunt commented:
I'm leaning towards not updating ipxe, we can make it use GNU-EFI 3. Currently rebasing my other PR that needed the new license from this.

lzcunt commented:
Hmm, it looks like ipxe does build on aarch64-linux on master with GNU-EFI 4. Maybe it was broken on master previously unrelated to this PR. I'm not making ipxe use GNU-EFI 3 then, however I haven't tested the resulting binaries.

JohnRTitor commented:
> I'm not making ipxe use GNU-EFI 3 then, however I haven't tested the resulting binaries.

There could be an underlying ABI change, which could break it like refind. So I would advise on pinning it unless we are sure that it works (from testing).
